tests/test_single_controller.py

```diff
@@ -0,0 +1,269 @@
+
+    def __init__(
+        self,
+        rank: int,
+        world_size: int,
+        master_addr: Optional[str] = None,
+        master_port: Optional[int] = None
+    ):
+        """Initialize the distributed GPU actor.
+
+        Args:
+            rank: The rank of this process in the distributed group
+            world_size: Total number of processes in the distributed group
+            master_addr: Master node address. If None, will allocate dynamically for rank 0
+            master_port: Master node port. If None, will allocate dynamically for rank 0
+        """
+        self.rank = rank
+        self.world_size = world_size
+        self.master_addr = master_addr
+        self.master_port = master_port
+
+        # Set up basic environment variables
+        os.environ['WORLD_SIZE'] = str(world_size)
+        os.environ['RANK'] = str(rank)
+
+        # Set LOCAL_RANK based on Ray GPU allocation
+        os.environ['LOCAL_RANK'] = '0' if is_cuda_visible_devices_set(
+        ) else str(ray.get_gpu_ids()[0])
+
+        # If this is rank 0 and no master_addr/master_port provided, allocate them
+        if rank == 0 and (master_addr is None or master_port is None):
+            self._allocate_master_address()
+
+        os.environ['MASTER_ADDR'] = self.master_addr
+        os.environ['MASTER_PORT'] = str(self.master_port)
+
+        self.model = None
+        self.model_update_group = None
+
+    def _allocate_master_address(self):
+        """Allocate master address and port for rank 0."""
+        if self.master_addr is None:
+            # Get the local IP address
+            self.master_addr = get_node_ip()
+
+        if self.master_port is None:
+            # Allocate a free port
+            self.master_port = get_free_port()
+
+    def get_master_address(self) -> tuple[Optional[str], Optional[int]]:
+        """Return the master address and port as a tuple."""
+        return (self.master_addr, self.master_port)
+
+    def get_free_port(self):
+        return get_free_port()
+
+    def init_train_process_group(self):
+        """Initialize the distributed process group."""
+        # Initialize process group
+        dist.init_process_group(timeout=timedelta(seconds=30))
+        logger.info(f'is distributed initialized: {dist.is_initialized()}')
+        # Print debug information
+        num_visible_devices = torch.cuda.device_count()
+        logger.info(f'num_visible_devices: {num_visible_devices}')
+        logger.info('Ray actor init envs:')
+        logger.info(f'rank: {dist.get_rank()}')
+        logger.info(f'node_rank: {dist.get_rank() // 8}')
+        logger.info(f'world_size: {dist.get_world_size()}')
+        logger.info(f'local_rank: {dist.get_rank() % 8}')
+        logger.info(f'master_addr: {self.master_addr}')
+        logger.info(f'master_port: {self.master_port}')
+
+    def init_model(self, model_name: str):
+        """Initialize the model."""
+        self.model = AutoModelForCausalLM.from_pretrained(
```

Comment:
i don't think you want torch_dtype auto, might end up with bf16 master weights. should always be float 32. i know this is just a test, just noting